### ISSUE TITLE
CRC32 hashing + filename check

### ISSUE THREAD

**Issue by darealshinji** [tedsmith/quickhash#28]
I still find videos on the Internet with CRC32 hashes in their filenames. Having a CRC32 check and letting the result compare with checksums found in their filenames would be useful to me.

**Comment by tedsmith** [comment]
Not too tricky to add, and the hash library supports CRC32. Do you have an example of some filenames, so I can see exactly what you mean?

**Comment by darealshinji** [comment]
Anime fansubs normally use them. Random example: `[Ginga-Subs]_Code_Geass_Akito_the_Exiled_01_[BD][1080p][34121F6F].mkv`

**Comment by tedsmith** [comment]
So the CRC32 hash of '[Ginga-Subs]_Code_Geass_Akito_the_Exiled_01_[BD][1080p][34121F6F].mkv' is 34121F6F, yes?

**Comment by darealshinji** [comment]
Exactly. Here's a tool that does such a thing, but it's very outdated: http://animechecker.sourceforge.net/

**Comment by darealshinji** [comment]
I'm currently working on a little tool that does this using zlib's CRC32 algorithm (`uLong crc32(uLong crc, const Bytef *buf, uInt len)`).
Here's the part where I convert the crc to upper and lower case strings and compare them with the filename:

``` C
  char *file = argv[1];
  long crc = crc32_zlib(file);
  if (crc == -1) {
    return 1;
  }

  char crc_upper[16];
  memset(crc_upper, '\0', 16);
  sprintf(crc_upper, "%.8lX", crc);

  char crc_lower[16];
  memset(crc_lower, '\0', 16);
  sprintf(crc_lower, "%.8lx", crc);

  printf("%s", crc_upper);

  if (strstr(file, crc_upper) != NULL || strstr(file, crc_lower) != NULL) {
    printf("\tOK");
  }
  printf("\n");
```
Checking the filename was actually easier than I thought. `strstr()` returns NULL if the strings aren't found. I assume there are similar functions in Pascal, so maybe that helps a little.

**Comment by darealshinji** [comment]
There's the package __hash__ which provides CRC32 functions: http://wiki.freepascal.org/hash
According to [this forum disussion](http://forum.lazarus.freepascal.org/index.php?topic=36396.0) source and example can be found in the FPC sources:
```
C:\Lazarus\fpc\3.0.2\source\packages\hash\src\crc.pas
C:\Lazarus\fpc\3.0.2\source\packages\hash\examples\crctest.pas
```

**Comment by tedsmith** [comment]
Thanks darealshinji. I've not forgot about this one...just been working on some other parts lately. Thanks for the pointers though if possible I'll stick with HashLib4Pascal...just keeps things simpler rather than using two libraries.

**Comment by darealshinji** [comment]
Checking for checksum in filename is simple:
``` pascal
// use LowerCase() if "sum" is lower case
if (AnsiContainsStr(UpperCase(filename), sum))
then
  result := ' OK  '
else
  result := ' --  ';
```

Here's a simple crc32 check program written in Pascal: https://gist.github.com/darealshinji/5913d311007754300b5a36281c3e470f
It's not using HashLib4Pascal though.

**Comment by tedsmith** [comment]
For sure. Technically the implementation isn't difficult and as you say CRC is an included Freepascal unit so we could easily use that or HashLib4Pascal. The bigger task is implementing it into the interface as an option, and then re-writing (adding to) all the loops of the various tabs of the program to implement the chosen hash (CRC in this case) if the user chooses that. And with all the SQLIte stuff I;ve been working on, I haven't got round to it yet. 

I'm guessing that you can use your own program to do those CRC computations when you need them for now, until I get round to implementing into QH?

**Comment by tedsmith** [comment]
Added to v3.3.0, pushed to branch but not merged to master yet. Will be available when v3.3.0 is released for all three platforms.
